packages/at_secondary_server/lib/src/verb/handler/pkam_verb_handler.dart

```diff
@@ -38,58 +39,122 @@ class PkamVerbHandler extends AbstractVerbHandler {
-          throw UnAuthenticatedException(response.errorMessage);
-        }
+      pkamAuthType = AuthType.apkam;
+      ApkamVerificationResult apkamResult =
+          await handleApkamVerification(enrollId, atSign);
+      if (apkamResult.response.isError) {
+        response.isError = apkamResult.response.isError;
+        response.errorCode = apkamResult.response.errorCode;
+        response.errorMessage = apkamResult.response.errorMessage;
+        return;
       }
+      publicKey = apkamResult.publicKey;
     } else {
+      pkamAuthType = AuthType.pkamLegacy;
       var publicKeyData = await keyStore.get(AT_PKAM_PUBLIC_KEY);
       publicKey = publicKeyData.data;
     }
 
-    // If there is no public key in the keystore then throw an exception
     if (publicKey == null || publicKey.isEmpty) {
-      response.data = 'failure';
-      response.isError = true;
-      response.errorMessage = 'pkam publickey not found';
       throw UnAuthenticatedException('pkam publickey not found');
     }
-    var isValidSignature = false;
+    var isValidSignature =
+        await _validateSignature(verbParams, sessionID, atSign, publicKey);
+    // authenticate if signature is valid
+    if (isValidSignature) {
+      atConnectionMetadata.isAuthenticated = true;
+      atConnectionMetadata.authType = pkamAuthType;
+      atConnectionMetadata.enrollmentId = enrollId;
+      response.data = 'success';
+    } else {
+      atConnectionMetadata.isAuthenticated = false;
+      logger.severe('pkam authentication failed');
+      throw UnAuthenticatedException('pkam authentication failed');
+    }
+  }
 
-    //retrieve stored secret using sessionid and atsign
-    var storedSecret = await keyStore.get('private:$sessionID$atSign');
-    storedSecret = storedSecret?.data;
+  @visibleForTesting
+  Future<ApkamVerificationResult> handleApkamVerification(
+      String enrollId, String atSign) async {
+    String enrollmentKey =
+        '$enrollId.$newEnrollmentKeyPattern.$enrollManageNamespace$atSign';
+    var enrollData = await keyStore.get(enrollmentKey);
+    final atData = enrollData.data;
+    logger.finer('fetched enrollData: $atData');
+    final enrollDataStoreValue =
+        EnrollDataStoreValue.fromJson(jsonDecode(atData));
+    ApkamVerificationResult apkamResult = ApkamVerificationResult();
+    Response response =
+        verifyEnrollApproval(enrollDataStoreValue.approval!.state, enrollId);
+    if (response.isError) {
+      apkamResult.response = response;
+      return apkamResult;
+    }
+    apkamResult.publicKey = enrollDataStoreValue.apkamPublicKey;
+    return apkamResult;
+  }
 
+  @visibleForTesting
+  Response verifyEnrollApproval(String approvalState, String enrollId) {
+    Response response = Response();
+    // the following is a function that based on the EnrollStatus sets
+    // appropriate error codes and messages
+    void getApprovalStatus(EnrollStatus enrollStatus) {
```

Comment:
Please do not have nested methods. Move getApprovalStatus outside verifyEnrollApproval method.

Reply:
addressed as part of [270b30a](https://github.com/atsign-foundation/at_server/pull/1520/commits/270b30a26cf30cd073e04d917877ec274009985e)